distributed/core.py

```diff
@@ -196,58 +196,42 @@ def handle_stream(self, stream, address):
 @gen.coroutine
 def read(stream, deserialize=True):
     """ Read a message from a stream """
-    if isinstance(stream, BatchedStream):
-        msg = yield stream.recv()
-        raise gen.Return(msg)
-    else:
-        n_frames = yield stream.read_bytes(8)
-        n_frames = struct.unpack('Q', n_frames)[0]
+    n_frames = yield stream.read_bytes(8)
+    n_frames = struct.unpack('Q', n_frames)[0]
 
-        lengths = yield stream.read_bytes(8 * n_frames)
-        lengths = struct.unpack('Q' * n_frames, lengths)
+    lengths = yield stream.read_bytes(8 * n_frames)
+    lengths = struct.unpack('Q' * n_frames, lengths)
 
-        frames = []
-        for length in lengths:
-            if length:
-                frame = yield stream.read_bytes(length)
-            else:
-                frame = b''
-            frames.append(frame)
+    frames = []
+    for length in lengths:
+        if length:
+            frame = yield stream.read_bytes(length)
+        else:
+            frame = b''
+        frames.append(frame)
 
-        msg = protocol.loads(frames, deserialize=deserialize)
-        raise gen.Return(msg)
+    msg = protocol.loads(frames, deserialize=deserialize)
+    raise gen.Return(msg)
 
 
 @gen.coroutine
 def write(stream, msg):
     """ Write a message to a stream """
-    if isinstance(stream, BatchedStream):
-        stream.send(msg)
-    else:
-        try:
-            frames = protocol.dumps(msg)
-        except Exception as e:
-            logger.info("Unserializable Message: %s", msg)
-            logger.exception(e)
-            raise
-
-        futures = []
-
-        lengths = ([struct.pack('Q', len(frames))] +
-                   [struct.pack('Q', len(frame)) for frame in frames])
-        futures.append(stream.write(b''.join(lengths)))
+    try:
+        frames = protocol.dumps(msg)
+    except Exception as e:
+        logger.info("Unserializable Message: %s", msg)
+        logger.exception(e)
+        raise
 
-        for frame in frames[:-1]:
-            futures.append(stream.write(frame))
+    lengths = ([struct.pack('Q', len(frames))] +
+               [struct.pack('Q', len(frame)) for frame in frames])
+    stream.write(b''.join(lengths))
 
-        futures.append(stream.write(frames[-1]))
+    for frame in frames[:-1]:
+        stream.write(frame)
 
-        while stream._write_buffer:
-            try:
-                yield gen.with_timeout(timedelta(seconds=0.01), futures[-1])
-                break
-            except gen.TimeoutError:
-                pass
+    yield stream.write(frames[-1])
```

Comment:
I *think* that this could fail if we write to the same stream in another coroutine.  Ideally we shouldn't do this.  Normally the `rpc` class creates new streams as necessary to handle concurrent communications to the same destination.  All cases that I can find when a coroutine writes directly to a stream it creates and owns that stream exclusively.

Still though, we were running into problems in the wild that suggested that this might be an issue.